src/java.base/share/classes/java/lang/String.java

```diff
@@ -217,11 +224,389 @@
+
+		if (COMPACT_STRINGS && (null == compressionFlag || (s1.coder | s2.coder | s3.coder) == LATIN1)) {
+			value = new byte[concatlen];
+			coder = LATIN1;
+
+			compressedArrayCopy(s1.value, 0, value, 0, s1len);
+			compressedArrayCopy(s2.value, 0, value, s1len, s2len);
+			compressedArrayCopy(s3.value, 0, value, s1len + s2len, s3len);
+		} else {
+			value = StringUTF16.newBytesFor(concatlen);
+			coder = UTF16;
+
+			// Check if the String is compressed
+			if (COMPACT_STRINGS && s1.coder == LATIN1) {
+				decompress(s1.value, 0, value, 0, s1len);
+			} else {
+				decompressedArrayCopy(s1.value, 0, value, 0, s1len);
+			}
+
+			// Check if the String is compressed
+			if (COMPACT_STRINGS && s2.coder == LATIN1) {
+				decompress(s2.value, 0, value, s1len, s2len);
+			} else {
+				decompressedArrayCopy(s2.value, 0, value, s1len, s2len);
+			}
+
+			// Check if the String is compressed
+			if (COMPACT_STRINGS && s3.coder == LATIN1) {
+				decompress(s3.value, 0, value, s1len + s2len, s3len);
+			} else {
+				decompressedArrayCopy(s3.value, 0, value, (s1len + s2len), s3len);
+			}
+
+			if (COMPACT_STRINGS) {
+				initCompressionFlag();
+			}
+		}
+	}
+
+	/*
+	 * Creates a string that is s1 + v1.
+	 */
+	private String(String s1, int v1) {
+		if (s1 == null) {
+			s1 = "null";
+		}
+
+		// Char length of all the parameters respectively
+		int s1len = s1.lengthInternal();
+		int v1len = 1;
+
+		int quot;
+		int i = v1;
+		while ((i /= 10) != 0)
+			v1len++;
+		if (v1 >= 0) {
+			quot = -v1;
+		} else {
+			// Leave room for '-'
+			v1len++;
+			quot = v1;
+		}
+
+		// Char length of the final String
+		int len = s1len + v1len;
+		if (len < 0) {
+			// K0D01 = Array capacity exceeded
+			throw new OutOfMemoryError(com.ibm.oti.util.Msg.getString("K0D01"));
+		}
+
+		if (COMPACT_STRINGS && (null == compressionFlag || s1.coder == LATIN1)) {
+			value = new byte[len];
+			coder = LATIN1;
+
+			// Copy in v1
+			int index = len - 1;
+
+			do {
+				int res = quot / 10;
+				int rem = quot - (res * 10);
+
+				quot = res;
+
+				// Write the digit into the correct position
+				helpers.putByteInArrayByIndex(value, index--, (byte) ('0' - rem));
+			} while (quot != 0);
+
+			if (v1 < 0) {
+				helpers.putByteInArrayByIndex(value, index, (byte) '-');
+			}
+
+			// Copy in s1 contents
+			compressedArrayCopy(s1.value, 0, value, 0, s1len);
+		} else {
+			value = StringUTF16.newBytesFor(len);
+			coder = UTF16;
+
+			// Copy in v1
+			int index = len - 1;
+
+			do {
+				int res = quot / 10;
+				int rem = quot - (res * 10);
+
+				quot = res;
+
+				// Write the digit into the correct position
+				helpers.putCharInArrayByIndex(value, index--, (char) ('0' - rem));
+			} while (quot != 0);
+
+			if (v1 < 0) {
+				helpers.putCharInArrayByIndex(value, index, (char) '-');
+			}
+
+			// Copy in s1 contents
+			decompressedArrayCopy(s1.value, 0, value, 0, s1len);
+
+			if (COMPACT_STRINGS) {
+				initCompressionFlag();
+			}
+		}
+	}
+
```

Comment:
I think we're missing a private constructor accelerated by the JIT:
https://github.com/eclipse/openj9/blob/a9838e74bba0f5935103bedcda6fc84c881804ca/jcl/src/java.base/share/classes/java/lang/String.java#L1156-L1337

Reply:
Good catch, just wondering if JIT falls back to a less-optimized call or we were just lucky not hit by `NoSuchMethodError`?

Reply:
We have an optimization which looks for String concatenations and replaces them with a call to a private String constructor. If that constructor is not found we just don't do the optimization (StringPeepholes).

Reply:
Filip, could you shine some lights on how the fallback works when this constructor [1] is not found? A quick experiment shows that missing a similar 3 parameter private constructor [2] caused build failure
```
Assertion failed at /Users/ivm/osx-openj9-builds/jdk17-0316-jdk17ristrV2/openj9-openjdk-jdk/omr/compiler/compile/OMRSymbolReferenceTable.cpp:979: method
VMState: 0x000535ff
	methodSymRefFromName: method must exist: java/lang/String.<init>(Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;)V
compiling java/util/Arrays.copyOfRange([BII)[B at level: warm
``` 

[1] https://github.com/eclipse/openj9/blob/a9838e74bba0f5935103bedcda6fc84c881804ca/jcl/src/java.base/share/classes/java/lang/String.java#L1156-L1159
[2] https://github.com/eclipse/openj9/blob/a9838e74bba0f5935103bedcda6fc84c881804ca/jcl/src/java.base/share/classes/java/lang/String.java#L1007-L1010

Reply:
This looks like the code which tries to find the symbol reference:
https://github.com/eclipse/openj9/blob/a9838e74bba0f5935103bedcda6fc84c881804ca/runtime/compiler/optimizer/StringPeepholes.cpp#L157-L214

I did not know that call to `methodSymRefFromName` will assert if it cannot find the symbol. So it looks like it is a bug not to have these methods.

Reply:
Added `private String(int v1, String s1, int v2, String s2, String s3)` for https://github.com/eclipse/openj9/blob/a9838e74bba0f5935103bedcda6fc84c881804ca/runtime/compiler/optimizer/StringPeepholes.cpp#L198 
Also added `String(byte[] data, int start, int length, boolean compressed)` for https://github.com/eclipse/openj9/blob/a9838e74bba0f5935103bedcda6fc84c881804ca/runtime/compiler/optimizer/StringPeepholes.cpp#L197

@fjeremic please have another look.